### PR TITLE
Add HTTP client and server tests

### LINKED ISSUE #6835: Add tests for netty library instrumentation
> Can you add some tests for the extracted libraries? (the usual `AbstractHttpClientTest`/`AbstractHttpServerTest` implementations)
This could be done in another PR -- I suspect it's going to take a couple hundred LOCs to implement that.

_Originally posted by @mateuszrzeszutek in https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6820#pullrequestreview-1134301488_

### PR DESCRIPTION
**Pull request by jack-berg**
Resolves #6835.

### CODE REVIEW

File: instrumentation/netty/netty-4.1/javaagent/src/test/groovy/Netty41ClientTest.groovy [hunk 1]
```diff
@@ -40,134 +33,18 @@ import java.util.concurrent.TimeUnit
 import static org.junit.jupiter.api.Assumptions.assumeTrue
 
 @Unroll
-class Netty41ClientTest extends HttpClientTest<DefaultFullHttpRequest> implements AgentTestTrait {
```
Comment (jack-berg) [inline instrumentation/netty/netty-4.1/javaagent/src/test/groovy/Netty41ClientTest.groovy:43]: I extracted as much of this to a `AbstractNetty41ClientTest` as I could but certain bits were tightly intwined with groove to easily pull apart.

File: instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyClientTelemetry.java [hunk 2]
```diff
@@ -63,4 +66,12 @@ public ChannelInboundHandlerAdapter createResponseHandler() {
       createCombinedHandler() {
     return new HttpClientTracingHandler(instrumenter);
   }
+
+  /**
+   * Propagate the current context to the {@link Channel}. This MUST be before HTTP request executed
+   * on a {@link Channel}.
```
Comment (mateuszrzeszutek) [inline instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyClientTelemetry.java:72]:
```suggestion
   * Propagate the current context to the {@link Channel}. This MUST be called before an HTTP request is executed
   * on a {@link Channel}.
```

File: instrumentation/netty/netty-4.1/testing/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/AbstractNetty41ServerTest.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+  static final LoggingHandler LOGGING_HANDLER =
+      new LoggingHandler(AbstractNetty41ServerTest.class, LogLevel.DEBUG);
+
+  protected abstract void configurePipeline(ChannelPipeline channelPipeline);
+
+  @Override
+  protected void configure(HttpServerTestOptions options) {
+    options.setTestException(false);
+    // etty instrumentation collects route but doesn't use it for span names
```
Comment (mateuszrzeszutek) [inline instrumentation/netty/netty-4.1/testing/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/AbstractNetty41ServerTest.java:51]:
```suggestion
    // netty instrumentation collects route but doesn't use it for span names
```
Does netty really collect `http.route`?

File: instrumentation/netty/netty-4.1/testing/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/AbstractNetty41ServerTest.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+            .handler(LOGGING_HANDLER)
+            .childHandler(
+                new ChannelInitializer<SocketChannel>() {
+                  @Override
+                  protected void initChannel(SocketChannel socketChannel) {
+                    ChannelPipeline pipeline = socketChannel.pipeline();
+                    pipeline.addFirst("logger", LOGGING_HANDLER);
+                    pipeline.addLast(new HttpServerCodec());
+                    pipeline.addLast(
+                        new SimpleChannelInboundHandler<Object>() {
```
Comment (mateuszrzeszutek) [inline instrumentation/netty/netty-4.1/testing/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/AbstractNetty41ServerTest.java:72]: Nit: WDYT of extracting this to a separate class? Just for better readability (less nesting/indentation)

File: instrumentation/netty/netty-4.1/javaagent/src/test/groovy/Netty41ClientTest.groovy [hunk 5]
```diff
@@ -204,8 +81,12 @@ class Netty41ClientTest extends HttpClientTest<DefaultFullHttpRequest> implement
           kind SpanKind.INTERNAL
           hasNoParent()
         }
-        clientSpan(it, 1, span(0))
-        serverSpan(it, 2, span(1))
+        span(1) {
+          kind SpanKind.CLIENT
+        }
```
Comment (mateuszrzeszutek) [inline instrumentation/netty/netty-4.1/javaagent/src/test/groovy/Netty41ClientTest.groovy:86]: Can you add `childOf` assertions to these?

**Review by mateuszrzeszutek — APPROVED** [review]
LGTM 👍 
Thanks!

File: instrumentation/netty/netty-4.1/testing/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/AbstractNetty41ServerTest.java (hunk 4 above)
Comment (jack-berg) [reply instrumentation/netty/netty-4.1/testing/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/AbstractNetty41ServerTest.java:72]: Yeah, I lifted this from a groovy test with minimal changes, but was thinking the same thing. Let's do a static inner class so its still encapsulated in the same class file.

File: instrumentation/netty/netty-4.1/testing/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/AbstractNetty41ServerTest.java (hunk 3 above)
Comment (jack-berg) [reply instrumentation/netty/netty-4.1/testing/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/AbstractNetty41ServerTest.java]: Nope! looks like I got confused trying to get the span name assertions to be correct given the interaction between HttpServerTestOptions.expectedHttpRoute, HttpServerTestOptions.expectedServerSpanNameMapper, and HttpServerTestOptions.httpAttributes.

File: instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyClientTelemetry.java [hunk 6]
```diff
@@ -63,4 +66,12 @@ public ChannelInboundHandlerAdapter createResponseHandler() {
...
     return new HttpClientTracingHandler(instrumenter);
   }
+
+  /**
+   * Propagate the current context to the {@link Channel}. This MUST be called before each HTTP
+   * request executed on a {@link Channel}.
+   */
+  public static void setChannelContext(Channel channel) {
+    channel.attr(AttributeKeys.WRITE_CONTEXT).compareAndSet(null, Context.current());
+  }
```
Comment (trask) [inline instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyClientTelemetry.java:76]: probably good to pass in `Context` as arg, for people who don't use thread-bound Context (e.g. some reactive apps)

also, then we could use this method from javaagent instrumentation too (e.g. reactor-netty)

**Review by trask — COMMENTED** [review]
nice 😅

File: instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyClientTelemetry.java (hunk 6 above)
Comment (jack-berg) [reply instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyClientTelemetry.java:76]: Ok, will start with one variant that accepts `Context` as an arg. Can add an overload where the context defaults to `Context.current()` later if needed.

**Review by trask — APPROVED** [review]

**Pull request merged**